# Prepare environment

In [ ]:
!pip install underthesea
!pip install jieba
!pip install gensim

     |████████████████████████████████| 7.5 MB 4.4 MB/s 
     |████████████████████████████████| 743 kB 59.7 MB/s 
     |████████████████████████████████| 3.1 MB 37.8 MB/s 
     |████████████████████████████████| 43 kB 1.8 MB/s 
     |████████████████████████████████| 235 kB 55.1 MB/s 
     |████████████████████████████████| 596 kB 46.9 MB/s 
     |████████████████████████████████| 61 kB 422 kB/s 
     |████████████████████████████████| 3.3 MB 26.4 MB/s 
     |████████████████████████████████| 895 kB 41.7 MB/s 
  Created wheel for seqeval: filename=seqeval-1.2.2-py3-none-any.whl size=16181 sha256=2587f4c4df19a6caec2c3687d5b24a712064cb680d5bc1c218f41ad74553baa5
  Stored in directory: /root/.cache/pip/wheels/05/96/ee/7cac4e74f3b19e3158dce26a20a1c86b3533c43ec72a549fd7
Successfully built seqeval
  Attempting uninstall: PyYAML
    Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13


In [ ]:
!pip install pyvi

     |████████████████████████████████| 8.5 MB 4.0 MB/s 


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity
from underthesea import word_tokenize, pos_tag, sent_tokenize
import warnings
from gensim import corpora, models, similarities
import jieba
import re


# Bước 1: Hiểu vấn đề, ngữ cảnh

### Ngữ cảnh
Giả sử doanh nghiệp thương mại điện tử top 2 Việt Name, tên là Tiki, chưa xây dựng hệ thống gợi ý sản phẩm tiêu dùng. Hiện Tiki muốn xây dựng hệ thống gợi ý này, để phát triển doanh nghiệp và cạnh tranh với các doanh nghiêp khác.

Và chúng ta là người được mời để phát triển hệ thống này.

### Xác định vấn đề
#### Mục tiêu:
Xây dựng hệ thống gợi ý sản phẩm tiêu dùng cho doanh nghiệp thương mại điện tử.
#### Giải pháp:
+ Content base recomendation (Gợi ý dựa trên sự tương đồng giữa các sản phẩm)
+ Collaborative filtering recomendation (Gợi ý dựa trên sự tương đồng và hành vi của các khách hàng với nhau)

Trong notebook này tập trung giải quyết bài toán đầu teine là content base recommendation

# Bước 2: Thu thập, hiểu dữ liệu

Dữ liệu đã được cấp sẵng với các thông tin như sau:

Dữ liệu được cung cấp sẵn gồm có các tập tin:
ProductRaw.csv, ReviewRaw.csv chứa thông tin sản phẩm,
review và rating cho các sản phẩm thuộc các nhóm hàng
hóa như Mobile_Tablet, TV_Audio, Laptop, Camera,
Accessory.

Gồm các cột như sau: 
+ item_id: mã sản phẩm
+ name: tên sản phẩm 
+ description: mô tả sản phẩm 
+ price: giá sản phẩm 
+ list_price: giá sản phẩm chưa khuyến mãi (nguyên bán)
+ rating: số sao đánh giá sản phẩm (1: tệ, đến 5: tốt)
+ brand: thương hiệu của sản phẩm
+ group: loại sản phẩm
+ url: đường link của sản phẩm
+ img: link hình ảnh của sản phẩm

In [ ]:
data = pd.read_csv('ProductRaw.csv')
data.head()

item_id  ...                                              image
0  48102821  ...  https://salt.tikicdn.com/cache/280x280/ts/prod...
1  52333193  ...  https://salt.tikicdn.com/cache/280x280/ts/prod...
2    299461  ...  https://salt.tikicdn.com/cache/280x280/media/c...
3  57440329  ...  https://salt.tikicdn.com/cache/280x280/ts/prod...
4  38458616  ...  https://salt.tikicdn.com/cache/280x280/ts/prod...

[5 rows x 10 columns]

In [ ]:
#Select useful columns for content base recomendation algothirm
data = data[['item_id', 'name', 'description', 'brand', 'group']]
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4404 entries, 0 to 4403
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   item_id      4404 non-null   int64 
 1   name         4404 non-null   object
 2   description  4401 non-null   object
 3   brand        4404 non-null   object
 4   group        4404 non-null   object
dtypes: int64(1), object(4)
memory usage: 172.2+ KB


In [ ]:
#Check duplicate data
len(data['item_id'].unique()) == data.shape[0]

False

In [ ]:
#Drop duplicate data
data = data.drop_duplicates(subset='item_id', keep = 'last')
len(data['item_id'].unique()) == data.shape[0]

True

In [ ]:
pd.options.display.max_rows = 250

In [ ]:
data['super_group'] = data['group'].apply(lambda x: x.split('-')[0])
data['super_group'].value_counts()

Thiết Bị Số                                             1065
Điện Tử                                                 1000
Máy Ảnh                                                  994
Laptop                                                   907
Điện Thoại                                               398
Hàng Quốc Tế/Laptop & Máy Vi Tính/Màn hình máy tính        2
Ô Tô                                                       2
Hàng Quốc Tế/Laptop & Máy Vi Tính/Linh kiện máy tính       2
Đồng hồ và Trang sức/Trang sức/Vòng tay                    1
Hàng Quốc Tế/Máy Ảnh                                       1
Nhà Cửa                                                    1
Name: super_group, dtype: int64

Đa số các sản phẩm đều là hàng tiêu dùng thiết bị số, công nghệ, ... Có 3 loại sản phẩm khác biệt hoàn toàn với các sản phẩm còn lại như: nhà cửa, đồng hồ và trang sức, ô tô
Loại sản phẩm chiếm số lượng nhiều nhất là máy ảnh, đồ dùng điện tử, thiết bị số, laptop.

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4373 entries, 0 to 4403
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   item_id      4373 non-null   int64 
 1   name         4373 non-null   object
 2   description  4370 non-null   object
 3   brand        4373 non-null   object
 4   group        4373 non-null   object
 5   super_group  4373 non-null   object
dtypes: int64(1), object(5)
memory usage: 239.1+ KB


Dữ liệu gồm 4373 dòng tương ứng với 4373 sản phẩm:  
+ không có dữ liệu thiếu
+ các cột đều đúng format
+ riêng cột object là kiểu chuỗi, thì sẽ phải nối các cột này lại với nhau để tạo nên feature raw_text. 

# Bước 3: Tiền xử lý dữ liệu
Vì những feature chưa thông tin quan trọng đều là dữ liệu dạng , do đó, bước làm sạch dữ liệu sẽ tập trung vào làm sạch dữ liệu dạng text, bao gồm các bước như sau:  

+ Tổng hợp các dữ liệu text thành một feature
+ Loại bỏ các ký tự không hợp lệ, stopwords, ...
+ Chuyển thành vector chuẩn hóa 


In [ ]:
#Create new feature
data['raw_text'] = data[[ 'name', 'description' ]].apply(lambda x: ' '.join([str(t) for t in x]), axis = 1)

data.head()

item_id  ...                                           raw_text
0  48102821  ...  Tai nghe Bluetooth Inpods 12 - Cảm biến vân ta...
1  52333193  ...  Tai nghe bluetooth không dây F9 True wireless ...
2    299461  ...  Chuột Không Dây Logitech M331 Silent - Hàng Ch...
3  57440329  ...  Loa Bluetooth 5.0 Kiêm Đồng Hồ Báo Thức - [[ 2...
4  38458616  ...  Tai Nghe Bluetooth Apple AirPods Pro True Wire...

[5 rows x 7 columns]

Dựa vào bài toán pos tag của underthesea: https://github.com/undertheseanlp/underthesea/wiki/M%C3%B4-t%E1%BA%A3-d%E1%BB%AF-li%E1%BB%87u-b%C3%A0i-to%C3%A1n-POS-Tag
ta có thể tạo preprocessor như sau:

+ Bước đầu tiên: lower, loại bỏ url, ký tự đặc biệt ra khỏi document
+ Bước hai: Tokenize document và gán nhãn loại từ cho từng token
+ Bước ba: Từ list các token và các nhãn, loại bỏ các stopwords, và các nhãn không quan trọng
+ Bước bốn: trả về kết quả là danh sách các từ đã được làm sạch

In [ ]:
#Create a preprocessor
class VietnamesePreprocessor():
  def __init__(self, additional_stopwords = None):
    from pyvi import ViTokenizer, ViPosTagger
    import re
    import numpy as np

    with open('vietnamese-stopwords.txt', 'r', encoding='utf-8') as f:
      self.stop_words = list(f.read().split('\n'))

    if additional_stopwords:
      if (type(additional_stopwords) =='str'):
        self.stop_words.append(additional_stopwords)
      else:
        self.stop_words.extend(additional_stopwords)

    self.stop_words = np.array(self.stop_words)
    self.tag_remove = ( 'C', 'F', 'I', 'L', 'M', 'Ny', 'Np', 'X', 'Nu', 'E') 
    self.remove_char = ['--', '//', '[[', '((', ']]', '))']
    self.nlp = lambda x: ViPosTagger.postagging(ViTokenizer.tokenize(x))
  
  def clean_text(self, text):
    import re
    text = re.sub('http://\S+|https://\S+', ' ', text)
    text = text.replace('\n', ' ') 

    #Remove default sentences on Tiki's description products.
    default_text1 = 'Giá sản phẩm trên Tiki đã bao gồm thuế theo luật hiện hành. Tuy nhiên tuỳ vào từng loại sản phẩm hoặc phương thức, địa chỉ giao hàng mà có thể phát sinh thêm chi phí khác như phí vận chuyển, phụ phí hàng cồng kềnh, ...'
    default_text2 = 'THÔNG TIN CHI TIẾT'
    default_text3 = 'MÔ TẢ SẢN PHẨM'
    text = text.replace(default_text1, ' ')
    text = text.replace(default_text2, ' ')
    text = text.replace(default_text3, ' ')
    return text.lower()

  def preprocess(self, document):

    document = self.clean_text(document)

    postag_ = self.nlp(document)

    bag_words = []

    for word, tag in zip(postag_[0], postag_[1]): 
      if word not in self.stop_words and tag not in self.tag_remove and len(word) > 1 and word not in self.remove_char:
        bag_words.append(word)

    return bag_words 



### Test class

In [ ]:
preprocessor = VietnamesePreprocessor(additional_stopwords=['công_nghệ', 'điện_tử', 'thiết_bị_số', 'điện', 'xuất_xứ', 'chi_tiết', 'thông_tin', 'thương_hiệu', 'màu_sắc', 'nâng_cấp',
                                                            'trọng_lượng', 'sản_phẩm', 'tỉ_lệ', 'nâng_cấp', 'thiết_kế', 'tiki', 'hàng', 'giao', 'phương_thức', 'chi_phí', 'phí',
                                                            'vận_chuyển', 'giá', 'thiết_bị', 'chất', 'thuế', 'luật'])

print(preprocessor.preprocess("Hôm nay tôi đi\nmua đồ ăn ở chợ, gặp ông bán cá rô phi\t, ổng nói tiếng việt, mua 100 con cá trả một cụt vàng"))
print(preprocessor.preprocess(data.iloc[0]['raw_text']))

['hôm_nay', 'mua', 'đồ', 'ăn_ở', 'chợ', 'cá_rô_phi', 'tiếng', 'mua', 'cá', 'cụt', 'vàng']
['tai_nghe', 'bluetooth', 'inpods', 'cảm_biến', 'vân', 'chống', 'đa_dạng', 'lựa_chọn', 'dung_lượng', 'pin', 'pin', 'nhạc', 'liên_tục', 'sạc', 'chờ', 'bluetooth', 'oem', 'trung', 'quốc', 'độ', 'nhạy_cảm_biến', 'vân', 'model', 'i12', 'jack', 'cắm', 'usb', 'cable', 'nhạc', 'liên_tục', 'sku', 'mô_tả', 'inpod', 'phiên_bản', 'tai_nghe', 'bluetooth', 'chuẩn', 'tai', 'airpod', 'hãng', 'lược', 'nút', 'bấm', 'thân', 'tai', 'thay', 'nút', 'cảm_ứng', 'dễ_dàng', 'thuận_tiện', 'thao_tác', 'cuộc_gọi', 'nhạc', 'dễ_dàng', 'chạm', 'bluetooth', 'kết_nối', 'vô_cùng', 'ổn_định', 'bluetooth', 'kết_nối', 'vô_cùng', 'ổn_định', 'tai_nghe', 'kết_nối', 'dock', 'sạc', 'âm', 'thời_lượng', 'pin', 'cải_thiện', 'tối_ưu', 'dock', 'sạc', 'tiện_lợi', 'sạc', 'pin', 'tai_nghe', 'tương_thích', 'bluetooth', 'xiaomi', 'samsung', 'apple', 'huawei', 'oppo', 'vivo', 'lenovo', 'tablet', 'nhạc', 'liên_tục', 'sạc', 'chờ', 'bao_gồm', 'hiện_hàn

# Bước 4: Tạo model

In [ ]:
#Create a similarity comparator

class VietnameseSimilarity():
  def __init__(self):
    pass

  def to_vectors(self, corpus):
      from sklearn.feature_extraction.text import TfidfVectorizer
      corpus = corpus.apply(lambda x: ' '.join(x))
      self.tfidf_model = TfidfVectorizer().fit(corpus)
      self.tfidf = self.tfidf_model.transform(corpus)
      return f'matrix sparse {self.tfidf.shape}'
  
  def to_similarity_matrix(self):
    self.similarityMatrix = cosine_similarity(self.tfidf, self.tfidf)
    return f'matrix sparse {self.similarityMatrix.shape}'

  def fit(self, corpus):
    self.to_vectors(corpus)
    return self.to_similarity_matrix()

  def compare(self, itemContent):
    from sklearn.metrics.pairwise import cosine_similarity
    new_vec =  self.tfidf_model.transform([itemContent])
    return cosine_similarity(new_vec, self.tfidf)[0]


In [ ]:
#Create a recommendator model
class ContentBaseRecommendator():
  def __init__(self):
    self.preprocessor = VietnamesePreprocessor(additional_stopwords=['công_nghệ', 'điện_tử', 'thiết_bị_số', 'điện', 'xuất_xứ', 'chi_tiết', 'thông_tin', 'thương_hiệu', 'màu_sắc', 'nâng_cấp',
                                                            'trọng_lượng', 'sản_phẩm', 'tỉ_lệ', 'nâng_cấp', 'thiết_kế', 'tiki', 'hàng', 'giao', 'phương_thức', 'chi_phí', 'phí',
                                                            'vận_chuyển', 'giá', 'thiết_bị', 'chất', 'thuế', 'luật'])
    self.comparator = VietnameseSimilarity()

  def fit(self, document, listItemsId):
    self.items = np.array(listItemsId)
    self.corpus = document.apply(self.preprocessor.preprocess)
    return self.comparator.fit(self.corpus)

  def recommend(self, itemContent, top_n = 10):
    itemContent = ' '.join(self.preprocessor.preprocess(itemContent))
    similarities = self.comparator.compare(itemContent)
    predData = pd.Series(data = similarities, index = self.items)
    predData = predData.where(predData < 1)
    return predData.nlargest(top_n)

    

  

In [ ]:
#Init model
model = ContentBaseRecommendator() 
#Fit to train data
model.fit(data['raw_text'], data['item_id'])


'matrix sparse (4373, 4373)'

In [ ]:
item_id = 48102821
item_description = data.loc[data['item_id'] == item_id, 'raw_text'][0]
item_description

'Tai nghe Bluetooth Inpods 12 - Cảm biến vân tay, chống nước,màu sắc đa dạng- 5 màu sắc lựa chọn THÔNG TIN CHI TIẾT\nDung lượng pin 300\nThời gian pin - Thời gian nghe nhạc liên tục từ 2.5-4h - Thời gian sạc đầy chỉ khoảng 60p - Thời gian chờ lên tới 140 giờ\nBluetooth 5\nThương hiệu OEM\nXuất xứ thương hiệu Trung Quốc\nĐộ nhạy cảm biến vân tay\nModel i12\nLoại Jack cắm USB Cable\nTrọng lượng 300g\nThời gian sử dụng\n- Thời gian nghe nhạc liên tục từ 2.5-4h\nSKU 4096608751631\nMÔ TẢ SẢN PHẨM\nINPOD 12 là phiên bản nâng cấp mới nhất , tai nghe Bluetooth 5.0 có thiết kế tỉ lệ chuẩn 1:1 với tai Airpod chính hãng\nLược bỏ nút bấm trên thân tai thay vào đó là nút cảm ứng sử dụng dễ dàng và thuận tiện hơn\nThao tác nhận cuộc gọi đến, chuyển nhạc...một cách dễ dàng chỉ bằng 1 chạm\nBluetooth 5.0 mới nhất cho kết nối vô cùng ổn định\n- Bluetooth 5.0 mới nhất cho kết nối vô cùng ổn định\n- 2 tai nghe tự kết nối với nhau khi được lấy ra từ dock sạc\n- Chất âm và thời lượng pin được cải thiện tối

In [ ]:
recommend_list = model.recommend(item_description)
recommend_list

56365197    0.535552
22413470    0.515882
48273751    0.505411
52333193    0.502786
56885678    0.492426
52889826    0.491312
50319688    0.490369
26348659    0.487378
35373097    0.482068
72928043    0.480268
dtype: float64

In [ ]:
data.loc[data['item_id'].isin(recommend_list.index), ['item_id','name']]

item_id                                               name
1    52333193  Tai nghe bluetooth không dây F9 True wireless ...
23   35373097  Tai Nghe Bluetooth True Wireless AMOI F9 5.0 C...
66   52889826  Tai Nghe Bluetooth Không Dây 5.0, AMOI F9, Cảm...
119  26348659  Tai Nghe Bluetooth True Wireless AMOI F9 5.0 C...
134  56885678  Tai Nghe Bluetooth TWS F9 Tai Nghe Nhét  Hai T...
147  22413470  Tai Nghe Bluetooth Air.podes Cảm Ứng Công Nghệ...
228  50319688  Tai Nghe Bluetooth Mini I12 Tws V5.0 (Trắng) N...
635  72928043  Tai Nghe Bluetooth Amoi F9 kèm Củ Sạc 1A và Cá...
639  56365197  Tai nghe bluetooth không dây i12 TWS 5.0, thiế...
719  48273751  Tai nghe Blutooth 5.0 kiêm dock sạc dự phòng- ...

#### **Bảng dưới đây là từ thư viện gensim**

In [ ]:
data.loc[data['item_id'].isin(recommend_list.index), ['item_id','name']]

item_id                                               name
1    52333193  Tai nghe bluetooth không dây F9 True wireless ...
23   35373097  Tai Nghe Bluetooth True Wireless AMOI F9 5.0 C...
30   35723184  Tai nghe Inpod i12 TWS Bluetooth 5.0 cho iPhon...
66   52889826  Tai Nghe Bluetooth Không Dây 5.0, AMOI F9, Cảm...
119  26348659  Tai Nghe Bluetooth True Wireless AMOI F9 5.0 C...
134  56885678  Tai Nghe Bluetooth TWS F9 Tai Nghe Nhét  Hai T...
147  22413470  Tai Nghe Bluetooth Air.podes Cảm Ứng Công Nghệ...
228  50319688  Tai Nghe Bluetooth Mini I12 Tws V5.0 (Trắng) N...
639  56365197  Tai nghe bluetooth không dây i12 TWS 5.0, thiế...
719  48273751  Tai nghe Blutooth 5.0 kiêm dock sạc dự phòng- ...

#### thử gợi ý sản phẩm bằng một content hoàn toàn mới so với data
link: https://tiki.vn/ban-phim-bosston-803-lot-p11242302.html?itm_campaign=CTP_YPD_TKA_PLA_UNK_ALL_UNK_UNK_UNK_UNK_X.84974_Y.757948_Z.1491128_CN.phim-803%252Blot&itm_medium=CPC&itm_source=tiki-ads&spid=37793205

In [ ]:
#predict new content
item_description = '''
Thông Số Kỹ Thuật: Màu sắc: Đen - Xám Chữ khắc Laser chống bay Kết nối: USB 2.0, tương thích mọi hệ điều hành Độ bền phím > 10.000.000 lượt bấm
Giới thiệu sản phẩm Bàn phím giả cơ Bosston 803 Led 7 màu Bàn phím led chuyên game giả cơ Bosston 803 với thiết kế tựa như 1 bàn phím cơ với đường nét và kiểu dáng rất phá cách, hệ thống đèn led 7 màu cực kỳ ấn tượng trãi đều trên mặt phím, giúp người dùng, đặc biệt là các game thủ có thể thoải mái sử dụng vào ban đêm. Điều đặc biệt ở bàn phím Bosston 803 là các nút bàn phím được thiết kế cao hơn so với các bàn phím thông thường khác, với khoảng cách từ mặt bàn phím lên đến mặt nút bấm là 7mm. Tạo một không gian trống bên dưới các nút phím giúp người dùng có cảm giác rất êm khi gõ, giống như đang sử dụng một bàn phím cơ thật sự. Ngoài ra với thiết kế phím cao như trên sẽ rất dễ dàng khi vệ sinh, hoặc khi vô tình đỗ nước vào cũng không bị ảnh hưởng các vi mạch bên trong. Giúp tránh những hư hỏng đáng tiếc có thể xảy ra với bàn phím. Thông Số Kỹ Thuật: - Màu sắc: Đen - Xám - Chữ khắc Laser chống bay - Kết nối: USB 2.0, tương thích mọi hệ điều hành - Độ bền phím > 10.000.000 lượt bấm - Kích thước : 445 * 130 * 35 mm - Đèn led nền và led phím với 7 màu bắt mắt. - Số Phím : 104 phím - Antighost bấm 1 lần được 19 phím - Phím giả cơ nghe âm thanh rất thanh và êm 
Giá sản phẩm trên Tiki đã bao gồm thuế theo luật hiện hành. Bên cạnh đó, tuỳ vào loại sản phẩm, hình thức và địa chỉ giao hàng mà có thể phát sinh thêm chi phí khác như phí vận chuyển, phụ phí hàng cồng kềnh, thuế nhập khẩu (đối với đơn hàng giao từ nước ngoài có giá trị trên 1 triệu đồng).....
'''
recommend_list = model.recommend(item_description)
recommend_list

10590207    0.774006
1044890     0.720605
4252075     0.720505
25732951    0.684765
11241055    0.682694
3953455     0.676614
588800      0.617189
16385902    0.593260
2883909     0.585814
11582921    0.525043
dtype: float64

In [ ]:
data.loc[data['item_id'].isin(recommend_list.index), ['item_id','name']]

item_id                                               name
89     4252075  Bộ Bàn Phím Giả Cơ và Chuột Chuyên Game G21 Le...
146   16385902           Bàn phím giả cơ chuyên game cao cấp G700
259    3953455  Bộ Bàn Phím Giả Cơ và Chuột Chuyên Game G21 Le...
266    1044890  Bàn Phím Giả Cơ Chuyên Game Bosston 803 LED (7...
599   25732951  Bàn phím giả cơ chuyên Game R8 1822 Led 7 màu ...
600   11241055                           Bàn Phím Giả Cơ G21 LED 
2607   2883909  Bộ Bàn Phím Chuột Có Dây Bosston 837 Tích Hợp ...
2730    588800  Bộ Bàn Phím Giả Cơ Và Chuột Chuyên Game R8 191...
2859  10590207  Bàn Phím Giả Cơ Chuyên Game R8 1822 LED 7 Màu ...
2866  11582921  [TẶNG LÓT CHUỘT] Bộ bàn phím giả cơ và chuột c...

In [ ]:
item_description = '''
Biden từ lâu đặt cược vào nỗ lực chấm dứt đại dịch để khẳng định di sản chính trị của mình, nhưng Omicron có thể trở thành gánh nặng mới.

"Giữa đại dịch, chính quyền Biden đã đưa người Mỹ đến trạng thái có thể thoải mái gặp gỡ gia đình, bạn bè, đi mua sắm, xem phim và làm những điều để đưa cuộc sống bình thường trở lại", Patti Solis Doyle, chiến lược gia đảng Dân chủ từng sát cánh cùng Joe Biden trong chiến dịch tranh cử tổng thống năm 2008, nói. "Tôi nghĩ ông ấy chưa được ghi nhận đúng mức về công lao đó".

Kể từ khi lên nắm quyền hồi tháng 1, Tổng thống Biden đang đứng trước rất nhiều khó khăn. Tỷ lệ ủng hộ ông liên tục giảm, một đồng minh thân cận thua trong cuộc bầu cử thống đốc bang Virginia, đảng Dân chủ đối mặt nguy cơ mất quyền kiểm soát lưỡng viện quốc hội sau cuộc bầu cử giữa nhiệm kỳ năm 2022.

Biden gần đây đã gặt hái một số thành công trong lĩnh vực lập pháp, khi luật cơ sở hạ tầng 1.200 tỷ USD được thông qua và dự luật Xây lại Tốt hơn trị giá 1,7 nghìn tỷ USD cũng nhiều khả năng sẽ vượt được ải quốc hội. Tuy nhiên, những thành công đó chưa được chuyển hóa thành kết quả tích cực trong các cuộc thăm dò dư luận.

Các cố vấn Nhà Trắng tin tưởng rằng tất cả những điều này sẽ thay đổi khi Covid-19 bị đẩy lùi. Niềm tin này dường như là lý do Biden đưa ra hai thông điệp khá mâu thuẫn khi biến chủng Omicron xuất hiện: cần cảnh giác cao với đại dịch, nhưng nên tiếp tục cuộc sống bình thường.

Sự tương phản này được thể hiện khá rõ nét khi Biden phát biểu về biến chủng Omicron hôm 29/11, xung quanh là những món đồ trang trí Giáng sinh ở Nhà Trắng. Hình ảnh này như lời nhắc nhở rằng mùa lễ hội đang đến rất gần, trong khi người Mỹ thấp thỏm chờ đợi thông tin rõ ràng hơn về mối đe dọa từ biến chủng mới.
'''
recommend_list = model.recommend(item_description)
recommend_list

50126505    0.115040
73922514    0.053773
58292845    0.053666
57447641    0.049716
59249458    0.047709
13446740    0.047343
25451926    0.043243
54935451    0.043220
13392137    0.043184
33651473    0.042857
dtype: float64

In [ ]:
data.loc[data['item_id'].isin(recommend_list.index), ['item_id','name']]

item_id                                               name
743   57447641  Bộ 2 bọc ngón tay cảm ứng chơi game mobile 2.0...
923   73922514  Tai Nghe Bluetooth True Wireless JBL Tune 115 ...
945   54935451  Đồng hồ thông minh cao cấp JAKCOM T500 màn hìn...
2628  13446740  Phần mềm diệt Virus Eset Internet Security 1 U...
2791  13392137  Phần mềm diệt Virus Eset Nod32 Antivirus 1 Use...
3463  25451926     Tivi LED Philips Full HD 40 inch 40PFT5063S/74
3473  59249458           Android Tivi Sony 4K 55 inch KD-55X9000H
3544  58292845           Android Tivi Sony 4K 55 inch KD-55X8050H
3738  33651473      Tivi LED Philips Full HD 40 inch 40PFT5583/74
4310  50126505   Micro Karaoke KTV Có Dây SR-688 (Màu Ngẫu Nhiên)

# Bước 5: Báo cáo kết quả

Hệ thống gợi ý sản phẩm bán lẻ trên tiki được xây dựng model như ở trên.

Hệ thống hoạt động dựa trên thuật toán: tfidf-cosine similarity của thư viện skitlearn, đưa ra gợi ý,sản phẩm tương đồng dựa trên tên và mô tả của sản phẩm.

Có thê thấy rõ, kết quả của hai thư viện sklearn và gesim là tương đồng nhau. Tuy nhiên về tốc độ thì gensim sẽ nhanh hơn, vì khi gợi ý content mới thì gensim không cần phải tính toán lại toàn bộ ma trận cosine similiarity nữa.

In [ ]:
#Save model
import dill as pickle
filename = '/content/gdrive/My Drive/LDS0_k271_HoangDucHuy/Project_03_Recommendation_1/model/sklearn.pkl'
pickle.dump(model, open(filename, 'wb'))

In [ ]:
#Test model
loaded_model = pickle.load(open(filename, 'rb'))
loaded_model.recommend(item_description)

50126505    0.115040
73922514    0.053773
58292845    0.053666
57447641    0.049716
59249458    0.047709
13446740    0.047343
25451926    0.043243
54935451    0.043220
13392137    0.043184
33651473    0.042857
dtype: float64